### PR TITLE
Add functions environment_set_variable() and filename_absolute()

### PR DESCRIPTION
**Pull request by time-killer-games**
```
// win32

// converts a relative path to absolute if the path exists
string filename_absolute(string fname) {
  if (file_exists(fname) || directory_exists(fname)) {
    wchar_t rpath[MAX_PATH];
    tstring tstr_fname = widen(fname);
    GetFullPathNameW(tstr_fname.c_str(), MAX_PATH, rpath, NULL); 
    return shorten(rpath);
  } else { return ""; }
}

// deletes the environment variable if set to empty string
bool environment_set_variable(const string &name, const string &value) {
  tstring tstr_name = widen(name);
  tstring tstr_value = widen(value);
  if (value == "") return (SetEnvironmentVariableW(tstr_name.c_str(), NULL) != 0);
  return (SetEnvironmentVariableW(tstr_name.c_str(), tstr_value.c_str()) != 0);
}
```
```
// posix

// converts a relative path to absolute if the path exists
string filename_absolute(string fname) {
  if (file_exists(fname) || directory_exists(fname)) {
    char rpath[PATH_MAX];
    char *ptr = realpath(fname.c_str(), rpath);
    if (ptr != NULL) { return rpath; }
  } return "";
}

// deletes the environment variable if set to empty string
bool environment_set_variable(const string &name, const string &value) {
  if (value == "") return (unsetenv(name.c_str()) == 0);
  return (setenv(name.c_str(), value.c_str(), 1) == 0);
}
```

### CODE REVIEW

**Comment by time-killer-games** [comment]
@RobertBColton this one is ready for you to review. It's a quicky.

**Comment by time-killer-games** [comment]
@JoshDreamland this one is ready for you to review. It's a quickie.

File: ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSmain.cpp [hunk 1]
```diff
@@ -457,14 +461,31 @@ void execute_program(std::string operation, std::string fname, std::string args,
...
 void execute_program(std::string fname, std::string args, bool wait) { execute_program("open", fname, args, wait); }
 
+// converts a relative path to absolute if the path exists
+std::string filename_absolute(std::string fname) {
+  if (file_exists(fname) || directory_exists(fname)) {
+    wchar_t rpath[MAX_PATH];
+    tstring tstr_fname = widen(fname);
+    GetFullPathNameW(tstr_fname.c_str(), MAX_PATH, rpath, NULL); 
+    return shorten(rpath);
+  } else { return ""; }
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSmain.cpp]: Generally, statements in braces get separated onto their own lines. In this case, the `else` isn't actually required. Also, it's inconsistent with your other method.

File: ENIGMAsystem/SHELL/Platforms/General/POSIX/POSIXdirectory.cpp [hunk 2]
```diff
@@ -25,9 +29,24 @@ bool set_working_directory(string dname) {
   return false;
 }
 
+// converts a relative path to absolute if the path exists
+std::string filename_absolute(std::string fname) {
+  if (file_exists(fname) || directory_exists(fname)) {
+    char rpath[PATH_MAX];
+    char *ptr = realpath(fname.c_str(), rpath);
+    if (ptr != NULL) { return rpath; }
+  } return "";
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Platforms/General/POSIX/POSIXdirectory.cpp]: I'd have put a newline here before return.

File: ENIGMAsystem/SHELL/Platforms/General/POSIX/POSIXdirectory.cpp [hunk 3]
```diff
@@ -25,9 +29,24 @@ bool set_working_directory(string dname) {
   return false;
 }
 
+// converts a relative path to absolute if the path exists
+std::string filename_absolute(std::string fname) {
+  if (file_exists(fname) || directory_exists(fname)) {
+    char rpath[PATH_MAX];
+    char *ptr = realpath(fname.c_str(), rpath);
+    if (ptr != NULL) { return rpath; }
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Platforms/General/POSIX/POSIXdirectory.cpp]: Probably better to return `ptr` in this case.

**Pull request merged**